### PR TITLE
[libaio] add new port

### PR DESCRIPTION
**Pull request by c8ef**
- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md).
- [x] The name of the port matches an existing name for this component on https://repology.org/ if possible, and/or is strongly associated with that component on search engines.
- [x] Optional dependencies are resolved in exactly one way. For example, if the component is built with CMake, all `find_package` calls are REQUIRED, are satisfied by `vcpkg.json`'s declared dependencies, or disabled with [CMAKE_DISABLE_FIND_PACKAGE_Xxx](https://cmake.org/cmake/help/latest/variable/CMAKE_DISABLE_FIND_PACKAGE_PackageName.html).
- [x] The versioning scheme in `vcpkg.json` matches what upstream says.
- [x] The license declaration in `vcpkg.json` matches what upstream says.
- [x] The installed as the "copyright" file matches what upstream says.
- [x] The source code of the component installed comes from an authoritative source.
- [x] The generated "usage text" is accurate. See [adding-usage](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/examples/adding-usage.md) for context.
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is in the new port's versions file.
- [x] Only one version is added to each modified port's versions file.

### CODE REVIEW

File: ports/libaio/portfile.cmake [hunk 1]
```diff
@@ -0,0 +1,19 @@
+vcpkg_download_distfile(ARCHIVE
+	URLS "https://pagure.io/libaio/archive/libaio-${VERSION}/libaio-libaio-${VERSION}.tar.gz"
+    FILENAME "libaio-${VERSION}.tar.gz"
```
Comment (MonicaLiu0311) [inline ports/libaio/portfile.cmake:3]:
```suggestion
    URLS "https://pagure.io/libaio/archive/libaio-${VERSION}/libaio-libaio-${VERSION}.tar.gz"
    FILENAME "libaio-${VERSION}.tar.gz"
```

File: ports/libaio/vcpkg.json [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+  "license": "LGPL-2.1",
+  "supports": "linux",
+  "dependencies": [
+    {
+      "name": "vcpkg-cmake",
+      "host": true
+    },
+    {
+      "name": "vcpkg-cmake-config",
+      "host": true
```
Comment (dg0yt) [inline ports/libaio/vcpkg.json:15]:
```suggestion
```
unused.

**Review by dg0yt — COMMENTED** [review]
On `CMakeLists.txt`:
- Make it an explicit file instead of a patch, and let the portfile copy it into `SOURCE_PATH`.
- Remove line breaks from the commands.
- Add double quotes around `${PROJECT_SOURCE_DIR}/src`
- Raise `cmake_minimum_required` to a more modern version.

**Review by MonicaLiu0311 — DISMISSED** [review]

**Comment by MonicaLiu0311** [comment]
The library provides neither `*config.cmake` file and `.pc` file. Could you please at least add a way to find library so that user will have clear `usage`?

**Review by dg0yt — CHANGES_REQUESTED** [review]
We know that some ports would use libaio. `cachelib` and `folly` have `libaio` features. These feature must depend on the new `libaio` port. This change needs to be part of this PR.

**Comment by dg0yt** [comment]
> The changes related to folly and cachelib will be in the following PR.

No.

This PR cannot be accepted without the changes to folly and cachelib: It would create immediate installation order problems with cached artifacts.

**Comment by dg0yt** [comment]
The problem is the following.
- Port `libaio` is installed first.
- Port `folly[libaio]` is installed second, not from cache.
  `folly[libaio]` will implicitly depend on the vcpkg package of `libaio` now. :bomb: 

Later:
- Port `libaio` is not installed.
- Port `folly[libaio]` is installed, from cache.
  The cached artifact of `folly[libaio]` implicitly depends on the vcpkg package of `libaio`, but this time it is missing (vcpkg) or mismatched (system) :boom:

**Comment by MonicaLiu0311** [comment]
When testing usage, the following error occurs:
```
monica@monica003:/mnt/test/build$ cmake --build .
[ 50%] Building CXX object CMakeFiles/main.dir/test.cpp.o
/mnt/test/test.cpp:2:10: fatal error: libaio.h: No such file or directory
    2 | #include "libaio.h"
      |          ^~~~~~~~~~
compilation terminated.
gmake[2]: *** [CMakeFiles/main.dir/build.make:76: CMakeFiles/main.dir/test.cpp.o] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:83: CMakeFiles/main.dir/all] Error 2
gmake: *** [Makefile:91: all] Error 2
```

<details>
<summary><b>test.cpp</b></summary>
<pre>
#include &#60;iostream&#62;
#include "libaio.h"

using namespace std;

int main()
{
        cout << "Hello CMake." << endl;
        return 0;
}
</pre></details>


<details>
<summary><b>CMakeLists.txt</b></summary>
<pre>
cmake_minimum_required (VERSION 3.8)

set(CMAKE_TOOLCHAIN_FILE "/mnt/aio/scripts/buildsystems/vcpkg.cmake")

project ("test")

add_library (main "test.cpp")

find_package(PkgConfig REQUIRED)
pkg_check_modules(LIBAIO REQUIRED IMPORTED_TARGET libaio)
target_include_directories(main PRIVATE ${LIBAIO _INCLUDE_DIRS})
target_link_libraries(main ${LIBAIO _LIBRARIES})
</pre></details>

**Comment by MonicaLiu0311** [comment]
The header files are found through the `.pc` file.

**Comment by dg0yt** [comment]
> `cachelib` and `folly` have `libaio` features. These feature must depend on the new `libaio` port. This change needs to be part of this PR.

AFAICS `cachelib` still needs to be done.

**Comment by dg0yt** [comment]
> The cachelib package does not directly procure the libaio package; instead, it acquires it through Folly.

I see.

File: ports/libaio/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,15 @@
+cmake_minimum_required(VERSION 3.20)
+project(libaio
+    VERSION 0.3.113
+    DESCRIPTION "The Linux-native asynchronous I/O facility"
+)
+file(GLOB SRC_FILES src/*.c)
+add_library(aio ${SRC_FILES})
+target_include_directories(aio PUBLIC "${PROJECT_SOURCE_DIR}/src")
+install(TARGETS aio LIBRARY DESTINATION lib)
```
Comment (dg0yt) [inline ports/libaio/CMakeLists.txt:9]:
```suggestion
install(TARGETS aio)
```
and use the defaults. (For `x64-linux`, it is not LIBRARY but ARCHIVE.)

File: ports/libaio/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+install(FILES src/libaio.h DESTINATION include)
+set(PKG_CONFIG_FILE_NAME "${CMAKE_CURRENT_BINARY_DIR}/${PROJECT_NAME}.pc")
+configure_file("${CMAKE_CURRENT_SOURCE_DIR}/pkgconfig.pc.in" "${PKG_CONFIG_FILE_NAME}" @ONLY)
+install(FILES "${PKG_CONFIG_FILE_NAME}"
+    DESTINATION "${CMAKE_INSTALL_LIBDIR}/pkgconfig"
+)
```
Comment (dg0yt) [inline ports/libaio/CMakeLists.txt:15]: FTR: This is not "official". Upstream doesn't provide a pc file.

**Review by dg0yt — COMMENTED** [review]
Nitpicks.

File: ports/libaio/portfile.cmake [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+vcpkg_extract_source_archive(
+    SOURCE_PATH
+    ARCHIVE "${ARCHIVE}"
+)
+
+file(COPY "${CMAKE_CURRENT_LIST_DIR}/CMakeLists.txt" DESTINATION "${SOURCE_PATH}")
+file(COPY "${CMAKE_CURRENT_LIST_DIR}/pkgconfig.pc.in" DESTINATION "${SOURCE_PATH}")
+
+vcpkg_cmake_configure(
+    SOURCE_PATH "${SOURCE_PATH}"
```
Comment (dg0yt) [inline ports/libaio/portfile.cmake:15]:
```suggestion
    SOURCE_PATH "${SOURCE_PATH}"
    OPTIONS
        -DVERSION=${VERSION}
```

File: ports/libaio/pkgconfig.pc.in [hunk 6]
```diff
@@ -0,0 +1,9 @@
+prefix=@CMAKE_INSTALL_PREFIX@
+includedir=${prefix}/include
+libdir=${prefix}/lib
+
+Name: @PROJECT_NAME@
+Description: @PROJECT_DESCRIPTION@
+Version: @PROJECT_VERSION@
```
Comment (dg0yt) [inline ports/libaio/pkgconfig.pc.in:7]:
```suggestion
Name: libaio
Description: The Linux-native asynchronous I/O facility
Version: @VERSION@

```

File: ports/libaio/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,15 @@
+cmake_minimum_required(VERSION 3.20)
+project(libaio
+    VERSION 0.3.113
+    DESCRIPTION "The Linux-native asynchronous I/O facility"
+)
```
Comment (dg0yt) [inline ports/libaio/CMakeLists.txt:5]:
```suggestion
project(libaio C)
```

**Review by dg0yt — CHANGES_REQUESTED** [review]
Don't hard code `PROJECT_VERSION` but pass through `VERSION`.
So we should not need to update `CMakeLists.txt` soon.
Use C only.

File: ports/libaio/CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+file(GLOB SRC_FILES src/*.c)
```
Comment (dg0yt) [inline ports/libaio/CMakeLists.txt:6]:
```suggestion

include(GNUInstallDirs)
file(GLOB SRC_FILES src/*.c)
```

File: ports/libaio/CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+add_library(aio ${SRC_FILES})
+target_include_directories(aio PUBLIC "${PROJECT_SOURCE_DIR}/src")
+install(TARGETS aio LIBRARY DESTINATION lib)
+install(FILES src/libaio.h DESTINATION include)
```
Comment (dg0yt) [inline ports/libaio/CMakeLists.txt:10]:
```suggestion
install(FILES src/libaio.h DESTINATION "${CMAKE_INSTALL_INCLUDEDIR}")

```

File: ports/libaio/pkgconfig.pc.in [hunk 10]
```diff
@@ -0,0 +1,9 @@
+prefix=@CMAKE_INSTALL_PREFIX@
+includedir=${prefix}/include
+libdir=${prefix}/lib
```
Comment (dg0yt) [inline ports/libaio/pkgconfig.pc.in:3]:
```suggestion
includedir=${prefix}/${CMAKE_INSTALL_INCLUDEDIR}
libdir=${prefix}/${CMAKE_INSTALL_LIBDIR}

```

**Review by dg0yt — COMMENTED** [review]
Consistent use of variables (and `GNUInstallDirs`) versus literal paths.

File: ports/libaio/pkgconfig.pc.in (hunk 10 above)
Comment (dg0yt) [reply ports/libaio/pkgconfig.pc.in]: Yeah, copy-and-paste failure :facepalm: 
I tested locally with `@` and `GNUInstallDirs`.

**Review by MonicaLiu0311 — DISMISSED** [review]

**Review by dg0yt — APPROVED** [review]
LGTM.

(I'm generally not happy with adding pc files which cannot be recognized as vcpkg-only, due to the potential lock-in effect. But there is no policy against this, and libaio.pc would be a likely choice for upstream.)

**Comment by BillyONeal** [comment]
vcpkg-team-review: Do we want to name the .pc file 'unofficial' or revisit the 'unofficial' policy overall since AFAICS we have never done the 'unofficial can be aliased to upstream targets' thing which was the original design intent?

**Comment by dg0yt** [comment]
> I offer an unofficial build system because I struggle to work with a Makefile-based system that lacks a configure script to manage `DESTDIR` or `INSTALL_PREFIX`. This package is straightforward, making it easy to create a CMake build for it.

There is no discussion about this.
The question is how to deal with exported usage. Upstream has no pc file, so there is no user for it until now (or until other evidence is found).
I know you were asked for "usage", but this could have been solved by documenting `find_path`, `find_library`.

**Comment by BillyONeal** [comment]
@AugP
@ras0219-msft
@data-queue
and I discussed this today.

@BillyONeal I think we should take this as is and also think we should reconsider the entire unofficial- policy.

@ras0219-msft I agree.

>(I'm generally not happy with adding pc files which cannot be recognized as vcpkg-only, due to the potential lock-in effect. But there is no policy against this, and libaio.pc would be a likely choice for upstream.)

We agree that creating lock-in is a problem. However, what we have observed in the past with the unofficial- policy is that it creates just as much, if not more, lock-in, because most of the time upstream takes what we have done or something very similar to what we have done, and then we're stuck patching everyone to add unofficial- everywhere. And customers get stuck with churn to get off the vcpkg-specific unofficial- thing.

**Review by BillyONeal — DISMISSED** [review]

**Comment by BillyONeal** [comment]
Thanks for the new port!

**Pull request merged**